### DISCUSSION TITLE
Better RTL Language Support

### DISCUSSION THREAD

**Discussion by bennypowers** [forem/forem#15471] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
Posts and comments written in RTL languages on dev.to don't always look great. See https://dev.to/biros/where-are-dev-users-coming-from-4l53/comments

**Describe the solution you'd like**
RTL posts and comments should be identified as such and have attributes or styles applied as needed. See https://github.com/waiting-for-dev/string-direction

**Describe alternatives you've considered**
Keen to hear other ideas.

**Additional context**
See also #304

**Comment by hummusonrails** [comment]
Solutions might be found within HTML5 itself, without the need for any additional Gems or packages. The following article outlines some steps that can be implemented: https://www.w3.org/International/questions/qa-html-dir. When the `dir` is set for RTL content, the punctuation marks are correctly placed at the end of a sentence and not moved to the front of a sentence, as an example.

**Comment by bennypowers** [comment]
@benhayehudi But how would we identify which content to apply `dir` to. Should we require authors who write in RTL languages to add some liquid or HTML tag to their markdown? Seems needlessly burdensome to me. I imagine we could run each tag in the rendered markdown through something like `string-direction` and conditionally apply a `dir` attribute should the need arise.

Would that significantly affect performance?
Could we reach a middle ground by adding a user setting like "scan my posts and comments for RTL languages"? What about discoverabililty - should we burden users even with a one-time config?

**Comment by N-Molham** [comment]
I just joined the community and would like to help if I can ... RTL support (Arabic) is pretty important

**Comment by hummusonrails** [comment]
@N-Molham - I agree. I think this is an important issue. I released an article in a RTL language, and the rendering was off because of it. I don't have a lot of spare time to code the solution, but glad to lend a hand in code review of it.

**Comment by KL13NT** [comment]
Hey people, is this issue going anywhere? I'd like to work on a PR for this but unsure whether I should even start. We could use a [CLD](https://github.com/CLD2Owners/cld2)-like library or simply ask the creator of an article to specify the direction of text. For instance, detecting/specfying the language and based on that add specific classes to the HTML that handle the directions and rendering of RTL text. This is quite needed right now.

**Comment by N-Molham** [comment]
I am not a python guy so I will try to help as much as I can

**Comment by cmgorton** [comment]
Hey folks we are starting work on i18n support on Forem. We are still in the initial infrastructure stages of this project. There is an open issue #14888 and a post by Ben Halpern that explains our plans a bit more: https://dev.to/devteam/forem-hacktoberfest-let-s-internationalize-404n

Because of this new work and the direction we are going with it, I am going to close this issue.

**Comment by yheuhtozr** [comment]
@cmgorton Unless you'd like to make #14888 an umbrella issue, I believe RTL support in web design is largely orthogonal to what that issue currently describes to do, and even workable alongside it. I wouldn't think closing this issue for the other makes sense in the i18n engineering context.

**Comment by citizen428** [comment]
@cmgorton I think I agree with @yheuhtozr on this, RTL-support in Forem will require a lot of work that goes beyond pure string translation.

**Comment by cmgorton** [comment]
OK! Thank you both for the input here. I am going to leave this closed for now while @JennieOcken and I work on the GitHub Discussions implementation. I think we could convert this issue to a discussion so we can continue talking about what would be needed. Then we could convert it back to an issue when we have a solid plan.

**Comment by ahangarha** [comment]
Hi
I wanted to share my thought on this issue. Instead of adding or improving RTL support, social media platforms like this need to have bidirectional text support. This means despite the global language and therefor direction of the page, the text in the opposite direction should be rendered in correct direction. This can be achieved by adding bidirectional text support.

In most of cases, a `dir=auto` does the trick but also there are some css modification needed.

I can help but I don't think I am technically in a position to make contribution on my own. At least I need one of you help me with setting up the project locally so that I can make the changes.

Please let me know if anyone can help me with this issue.

**Comment by yonimiz** [comment]
Hi
Coming back to this issue, I see it is holding for a long time.
Looks like the only issue is the right-to-left text alignment, is there any progress or plan for this?
LTR will help both Arabic and Hebrew languages.

**Comment by ahangarha** [comment]
@yonimiz 
Please keep pinging me on this issue. I know the solution, and I just need to find a chance to sit and implement it.

As mentioned before, we need bidirectional text support so the user-generated content gets rendered in the proper direction without explicit direction mentioned. Alignment is another thing. It is related but also different.

Mastodon and GitLab already do this. We should push for more bidirectional text support on the web.
